Userland/Libraries/LibWeb/WebAudio/AnalyserNode.cpp

```diff
@@ -0,0 +1,364 @@
+    // the previously computed data is returned.
+    //      Need to implement some kind of blocking mechanism, I guess
+    //      Might be more obvious how to handle this when render quantua have some
+    //      more scaffolding
+    //
+
+    // current_frequency_data returns a vector of size m_fftSize
+    // FIXME: Ensure sizes are correct after the fourier transform is implemented
+    //        Spec says to write frequencyBinCount bytes, not fftSize
+    Vector<f32> dB_data = current_frequency_data();
+    Vector<u8> byte_data;
+    byte_data.ensure_capacity(dB_data.size());
+
+    // For getByteFrequencyData(), the 𝑌[𝑘] is clipped to lie between minDecibels and maxDecibels...
+    // FIXME: Naive
+    for (auto x : dB_data) {
+        if (x < m_min_decibels)
+            x = m_min_decibels;
+
+        if (x > m_max_decibels)
+            x = m_max_decibels;
+    }
+
+    // ... and then scaled to fit in an unsigned byte such that minDecibels is represented by the
+    // value 0 and maxDecibels is represented by the value 255.
+    f32 delta_dB = m_max_decibels - m_min_decibels;
+    for (size_t i = 0; i < dB_data.size(); i++)
+        byte_data[i] = u8(255 * (dB_data[i] - m_min_decibels) / delta_dB);
+
+    // Write the current frequency data into array. If array’s byte length is less than frequencyBinCount,
+    // the excess elements will be dropped. If array’s byte length is greater than the frequencyBinCount ,
+    // the excess elements will be ignored. The most recent fftSize frames are used in computing the frequency data.
+    auto& output_buffer = array->viewed_array_buffer()->buffer();
+    size_t bytes_to_write = min(array->byte_length(), frequency_bin_count());
+
+    for (size_t i = 0; i < bytes_to_write; i++)
+        output_buffer[i] = byte_data[i];
+
+    return {};
+}
+
+// https://webaudio.github.io/web-audio-api/#dom-analysernode-getfloattimedomaindata
+WebIDL::ExceptionOr<void> AnalyserNode::get_float_time_domain_data(JS::Handle<WebIDL::BufferSource> const& array)
+{
+    // Write the current time-domain data (waveform data) into array. If array has fewer elements than the
+    // value of fftSize, the excess elements will be dropped. If array has more elements than the value of
+    // fftSize, the excess elements will be ignored. The most recent fftSize frames are written (after downmixing).
+
+    Vector<f32> time_domain_data = current_time_domain_data();
+
+    auto& vm = this->vm();
+
+    if (!is<JS::Float32Array>(*array->raw_object()))
+        return vm.throw_completion<JS::TypeError>(JS::ErrorType::NotAnObjectOfType, "Float32Array");
+    auto& output_array = static_cast<JS::Float32Array&>(*array->raw_object());
+
+    size_t floats_to_write = min(output_array.data().size(), frequency_bin_count());
+    for (size_t i = 0; i < floats_to_write; i++) {
+        output_array.data()[i] = time_domain_data[i];
+    }
```

Comment:
Along with adding tests, copying into the `JS::*_Array`s is the last sketchy thing here. The math is done in `AK::Vector`s, and it's not immediately obvious to me how to do a copy from them into arrays